pkg/controller/gkenetworkparamset/gkenetworkparamset_controller.go

```diff
@@ -174,27 +209,93 @@ func (c *Controller) syncGKENetworkParamSet(ctx context.Context, key string) err
 	}
 
 	params := obj.(*networkv1.GKENetworkParamSet)
+	originalParams := params.DeepCopy()
 
-	// TODO: Enable finalizer addition when finalizer deletion is added.
-	// addFinalizerToGKENetworkParamSet(params)
-	// update will be done once in deferred function call
-	// if err := c.updateGKENetworkParamSet(ctx, params); err != nil {
-	// 	return err
-	// }
+	defer func() {
```

Comment:
why do you add this global defers? they are hard to read and prone to introduce subtle bugs on refactors, is not linear the function, I see multiple returns after checking the errors, do you want to patch despite something failed?

Reply:
Yes we **always** want to update the object if its modified at any point in the sync function. If there is an error midway through the sync function we'd still like to cause the update to report back conditions that have already been validated.

Getting rid of the defer means we need to constantly call the update function after each transformation to GNP or network.


As for reflect.DeepEqual performance, this sync loop is not going to be run often enough for it to be an issue. Customers are limited in how many Networks and GNP objects they can create. It simplifies the code because we don't need to keep track if a change was made in every transformation function that needs to be propagated to this deferred update.

If it becomes a performance issue later on, we can always go back and change this according to your suggestion.